File: src/main/java/com/wesker926/leetcode/algorithms/p0124/s1/solution.go
// Package s1
// @author: wesker.gh
// @date: 2022/2/16
// @description: DFS
package s1

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

var ans, max int

func maxPathSum(root *TreeNode) int {
	ans, max = 0, root.Val
	dfs(root)
	if ans == 0 {
		return max
	}
	return ans
}

func dfs(node *TreeNode) int {
	if node == nil {
		return 0
	}

	max = maxInt(max, node.Val)
	l, r := dfs(node.Left), dfs(node.Right)
	res := maxInt(maxInt(maxInt(l, r), 0)+node.Val, 0)
	ans = maxInt(ans, maxInt(res, l+node.Val+r))
	return res
}

func maxInt(i, j int) int {
	if i >= j {
		return i
	}
	return j
}

File: src/main/java/com/wesker926/leetcode/algorithms/p0124/s1/Solution.java
package com.wesker926.leetcode.algorithms.p0124.s1;

/**
 * @author wesker.gh
 * @date 2022/5/20
 * @description DFS
 */
public class Solution {
    public int maxPathSum(TreeNode root) {
        int[] ans = new int[]{root.val};
        dfs(root, ans);
        return ans[0];
    }

    private int dfs(TreeNode node, int[] ans) {
        if (node == null) {
            return 0;
        }
        int l = dfs(node.left, ans), r = dfs(node.right, ans);
        ans[0] = Math.max(ans[0], l + r + node.val);
        return Math.max(Math.max(l, r) + node.val, 0);
    }

    private static class TreeNode {

        int val;

        TreeNode left;

        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
}
